### PR TITLE
Allow specification of multiple expanders

### PR DESCRIPTION
**Pull request by ryanmcnamara**
Multiple expanders can now be specified, expanders now "filter to the
tied for best" instead of "selecting the best" so the output of one
expander is now fed to the input of the next. Each expander may only
be used once to disallow bad configuration. This should not be a change
in functionality as in the event of a tie the random expander is still
used.

### CODE REVIEW

**Comment by ryanmcnamara** [comment]
fixes https://github.com/kubernetes/autoscaler/issues/4206

**Comment by ryanmcnamara** [comment]
Built and deployed this locally and confirmed I was able to use `--expander=priority,least-waste` and it worked as expected

File: cluster-autoscaler/expander/factory/expander_factory.go
Comment (x13n) [inline cluster-autoscaler/expander/factory/expander_factory.go:48]: Shouldn't it be an error to specify an expander that's guaranteed to provide a single option with a fallback? E.g. `random,least-waste` is not really a reasonable config.

File: cluster-autoscaler/expander/factory/chain.go
Comment (x13n) [inline cluster-autoscaler/expander/factory/chain.go:24]: There is non-trivial logic behind BestOptions call sometime, so maybe worth returning earlier if there's just one option left?

Comment (ryanmcnamara) [reply cluster-autoscaler/expander/factory/chain.go:24]: +1, will do

File: cluster-autoscaler/expander/factory/expander_factory.go
Comment (ryanmcnamara) [reply cluster-autoscaler/expander/factory/expander_factory.go:48]: True that the configuration is not a reasonable one, but I felt that catching this case wasn't worth the effort as it would require disambiguating between expanders that

1. Always return one result and
2. Those that can return more than one.

My thinking is that since it won't create problems then we can keep the simplicity of allowing this. Happy to be overruled on this though.

Comment (ryanmcnamara) [reply cluster-autoscaler/expander/factory/expander_factory.go:48]: Here's what it would look like: https://github.com/ryanmcnamara/autoscaler/compare/rm/expander-chain...ryanmcnamara:rm/expander-chain-demo?expand=1

File: cluster-autoscaler/expander/factory/chain.go
Comment (ryanmcnamara) [reply cluster-autoscaler/expander/factory/chain.go:24]: done

File: cluster-autoscaler/expander/factory/expander_factory.go
Comment (x13n) [reply cluster-autoscaler/expander/factory/expander_factory.go:48]: Yup, something like this. I'm not insisting on this one, but since you're generating an error for one class of invalid configs (duplicated entries), I thought it may make sense to go a step further and disallow bad ordering as well.

Comment (ryanmcnamara) [reply cluster-autoscaler/expander/factory/expander_factory.go:52]: Ok addressed in this pr now!

**Comment by ryanmcnamara** [comment]
Looking for a review from a code owner, perhaps @MaciekPytel or if not can you redirect? Thanks!

File: cluster-autoscaler/expander/factory/expander_factory.go
Comment (x13n) [reply cluster-autoscaler/expander/factory/expander_factory.go:52]: Neat, thanks! Maybe worth adding a comment about this behavior? May be surprising for anyone adding a new expander in the future. Please also squash commits before merging.

**Comment by x13n** [comment]
I'm about to go on vacation tomorrow, so to make sure it's not blocked on me in the meantime:

/lgtm
/hold until the commits are squashed
/assign @MaciekPytel for final approval

**Comment by x13n** [comment]
Just realized this is still open. Can you squash the commits?

**Comment by ryanmcnamara** [comment]
@x13n done!

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Comment by mwielgus** [comment]
Please fix:
```
Boilerplate header is wrong for: /home/runner/work/autoscaler/autoscaler/go/src/k8s.io/autoscaler/cluster-autoscaler/expander/factory/chain_test.go
```

**Comment by ryanmcnamara** [comment]
@mwielgus done, thanks for the review

**Comment by jonathanbbai** [comment]
I don't want to upgrade my cluster to 1.23 but i really like this feature, You think if I upgrade the CA version to 1.23 while on a 1.20 cluster, what are the chances it'll work?

**Pull request merged**
